### PR TITLE
Add OpenStack credential propagation support

### PR DESCRIPTION
**Pull request by bnallapeta**
Local Run:

```
ka get machines
NAME                           CLUSTER         NODENAME                       PROVIDERID                                          PHASE     AGE     VERSION
openstack-dev-cp-0             openstack-dev   openstack-dev-cp-0             openstack:///b52ccc71-566b-4487-a0b8-64b2c08f0e5b   Running   8m29s   v1.30.5+k0s.0
openstack-dev-cp-1             openstack-dev   openstack-dev-cp-1             openstack:///4022efcb-da46-4a99-80f3-2eaf3681238e   Running   8m29s   v1.30.5+k0s.0
openstack-dev-cp-2             openstack-dev   openstack-dev-cp-2             openstack:///87ae87e0-2ac5-4f06-b7ba-970d6dbcab3f   Running   8m29s   v1.30.5+k0s.0
openstack-dev-md-p27j6-6wfn4   openstack-dev   openstack-dev-md-p27j6-6wfn4   openstack:///5c2c51c8-f2c0-48d3-9c8f-ed28db53112a   Running   8m30s   
openstack-dev-md-p27j6-w6qvx   openstack-dev   openstack-dev-md-p27j6-w6qvx   openstack:///2df39a6d-2507-452f-9a0a-fb3e6a5d2052   Running   8m30s   
```

```
clusterctl describe cluster openstack-dev -n hmc-system
NAME                                                      READY  SEVERITY  REASON  SINCE  MESSAGE                                                        
Cluster/openstack-dev                                     True                     5m26s                                                                  
├─ClusterInfrastructure - OpenStackCluster/openstack-dev                                                                                                  
├─ControlPlane - K0sControlPlane/openstack-dev-cp                                                                                                         
│ └─3 Machines...                                         True                     7m32s  See openstack-dev-cp-0, openstack-dev-cp-1, ...                 
└─Workers                                                                                                                                                 
  └─MachineDeployment/openstack-dev-md                    True                     110s                                                                   
    └─2 Machines...                                       True                     3m43s  See openstack-dev-md-p27j6-6wfn4, openstack-dev-md-p27j6-w6qvx  
```

```
k -n hmc-system get managedclusters.hmc.mirantis.com 
NAME            READY   STATUS
openstack-dev   True    ManagedCluster is ready
```

### CODE REVIEW

File: internal/credspropagation/openstack.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+	// Apply CCM config
+	if err := applyCCMConfigs(ctx, cfg.KubeconfSecret, ccmSecret); err != nil {
+		return fmt.Errorf("failed to apply OpenStack CCM secret: %w", err)
+	}
+
+	return nil
+}
+
+func generateOpenStackCCMSecret(openstackSecret *corev1.Secret) *corev1.Secret {
```
Comment (eromanova) [inline internal/credspropagation/openstack.go:66]: You should create the secret with OpenStack cloud configuration following the specific format. It's not the same as the secret with credentials in the management cluster (the format: https://github.com/kubernetes/cloud-provider-openstack/blob/master/manifests/controller-manager/cloud-config).

Related thing: credentials should not be exposed, so instead of passing the credentials as chart values you should pass the name of this secret to the CCM and CSI values: https://github.com/Mirantis/hmc/pull/696/files#diff-c5afee6dd916ada2ea759da67744b5845149af9a6532b3891574865168c2c44e
See: https://github.com/kubernetes/cloud-provider-openstack/blob/master/charts/cinder-csi-plugin/values.yaml#L162

Similarly, it's done for the azure and the vsphere providers. You can use it as a reference.

**Review by eromanova — CHANGES_REQUESTED** [review]

File: internal/credspropagation/openstack.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+		return fmt.Errorf("failed to get ManagedCluster %s: %w", cfg.ManagedCluster.Name, err)
+	}
+
+	openstackCredential := &hmc.Credential{}
+	if err := cfg.Client.Get(ctx, client.ObjectKey{
+		Name:      openstackManagedCluster.Spec.Credential,
+		Namespace: openstackManagedCluster.Namespace,
+	}, openstackCredential); err != nil {
+		return fmt.Errorf("failed to get OpenStackCredential %s: %w", cfg.ManagedCluster.Spec.Credential, err)
+	}
```
Comment (a13x5) [inline internal/credspropagation/openstack.go:42]: This call could be avoided. We already requested this object before in the managedcluster_controller.

File: internal/credspropagation/openstack.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+	// Fetch the secret containing OpenStack credentials
+	openstackSecret := &corev1.Secret{}
+	openstackSecretName := openstackCredential.Spec.IdentityRef.Name
+	openstackSecretNamespace := openstackCredential.Spec.IdentityRef.Namespace
```
Comment (a13x5) [inline internal/credspropagation/openstack.go:47]: These aren't needed - you use this variables just once, consider passing values directly into struct below.

**Review by a13x5 — CHANGES_REQUESTED** [review]
Several issues with overall logic.
Also I believe the scope of this PR could be broaden a little bit by adding credential passing to the cluster (`setIdentityHelmValues` function). Currently we're expecting ClusterIdentity object only, this isn't the case with OpenStack provider, so logic should be adjusted/tested. If you prepared another PR with these changes it should be merged first.

File: internal/credspropagation/common.go [hunk 4]
```diff
@@ -53,11 +53,11 @@ func applyCCMConfigs(ctx context.Context, kubeconfSecret *corev1.Secret, objects
 	return nil
 }
 
-func makeSecret(name, namespace string, data map[string][]byte) *corev1.Secret {
+func makeSecret(name string, data map[string][]byte) *corev1.Secret {
```
[the remark this answers is not in the document]
Comment (bnallapeta) [reply internal/credspropagation/common.go:56]: https://github.com/Mirantis/hmc/actions/runs/12067350380/job/33650115071

This is the reason for this change. I am not sure how it passed lint for other functions which I have updated too. Basically it would always get `metav1.NamespaceSystem` which would always evaluate to `kube-system`.

I think when we get to that point of changing the namespace, we can parameterize this.

File: internal/credspropagation/openstack.go (hunk 3 above)
Comment (bnallapeta) [reply internal/credspropagation/openstack.go:47]: Why? Its generally a good practice to use variables. What if its used just once?

File: internal/credspropagation/openstack.go (hunk 2 above)
Comment (bnallapeta) [reply internal/credspropagation/openstack.go:42]: Not that a single api call would impact api server anyways. Especially since we would have 1 Credential resource. But sure, have removed this and passed params from managedcluster_controller.go

File: internal/credspropagation/openstack.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+)
+
+func PropagateOpenStackSecrets(ctx context.Context, cfg *PropagationCfg) error {
+	openstackManagedCluster := &hmc.ManagedCluster{}
+	if err := cfg.Client.Get(ctx, client.ObjectKey{
+		Name:      cfg.ManagedCluster.Name,
+		Namespace: cfg.ManagedCluster.Namespace,
+	}, openstackManagedCluster); err != nil {
+		return fmt.Errorf("failed to get ManagedCluster %s: %w", cfg.ManagedCluster.Name, err)
+	}
```
[the remark this answers is not in the document]
Comment (bnallapeta) [reply internal/credspropagation/openstack.go:34]: Removed.

File: internal/credspropagation/openstack.go (hunk 1 above)
Comment (bnallapeta) [reply internal/credspropagation/openstack.go:50]: @eromanova Yes, it already follows the same format. And based on the other comments, I have removed passing values to ccm and csi and am passing just the secret. So, this should ideally work. 

@a13x5 Yes, I defined a function here with the intention of doing the INI format thing. But later realized I don't have to do that. I agree that we don't need this function. Modified.


I will anyways test this out and make sure everything works.

File: internal/credspropagation/common.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (a13x5) [reply internal/credspropagation/common.go:56]: Well, shit...
Got it, then please adjust `makeConfigMap` as well, so both functions have the same signature. 
It would be good also to rename them to reflect that it will make secret in kube ns, but it's up to you really.

File: internal/credspropagation/openstack.go (hunk 2 above)
Comment (a13x5) [reply internal/credspropagation/openstack.go]: It's not just single API call here and it's not depends on the amount of `Credential` resource, but rather on amount of `ManagedCluster` resources. 
So it's more like one more call for each reconcile for each `ManagedCluster`.

File: internal/credspropagation/openstack.go (hunk 3 above)
Comment (a13x5) [reply internal/credspropagation/openstack.go]: Generally yes.
But in this particular example you already have variable `openstackCredential`. You copied part of it in other vars without any processing to pass them as a parameters and discard.
It would make sense to use variables here in case of:
- modifying data from `openstackCredential`
- using it in multiple places to improve readability (doubtful, but ok)
TL;DR here it just looks weird.

Comment (bnallapeta) [reply internal/credspropagation/openstack.go]: :)

File: internal/credspropagation/common.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (bnallapeta) [reply internal/credspropagation/common.go:56]: Have updated makeConfigMap as well. But haven't renamed it to leave it flexible for future. (if/when we add a namespace param)

**Comment by bnallapeta** [comment]
@a13x5 @eromanova Would appreciate if you can take a look at the PR and merge it soon.

File: config/dev/openstack-credentials.yaml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+stringData:
+  clouds.yaml: |
+    clouds:
+      openstack:
+        auth:
+          auth_url: ${OS_AUTH_URL}
+          username: ${OS_USERNAME}
+          password: ${OS_PASSWORD}
+          project_id: ${OS_PROJECT_ID}
+          project_name: ${OS_PROJECT_NAME}
```
Comment (eromanova) [inline config/dev/openstack-credentials.yaml:16]: AFAIR, the `project_name` is optional. Will this config be properly rendered if the `${OS_PROJECT_NAME}` is unset?

**Review by eromanova — COMMENTED** [review]
Small comment from my side. The rest looks good.

**Review by a13x5 — APPROVED** [review]
Approved
Additional fixes will be made later

File: internal/credspropagation/openstack.go [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+	ApplicationCredentialID     string `yaml:"application_credential_id"`
+	ApplicationCredentialName   string `yaml:"application_credential_name"`
+	ApplicationCredentialSecret string `yaml:"application_credential_secret"`
+	Username                    string `yaml:"username"`
+	Password                    string `yaml:"password"`
+	ProjectID                   string `yaml:"project_id"`
+	ProjectName                 string `yaml:"project_name"`
+	UserDomainName              string `yaml:"user_domain_name"`
+	ProjectDomainName           string `yaml:"project_domain_name"`
+}
```
Comment (zerospiel) [inline internal/credspropagation/openstack.go:72]:
```suggestion
type (
	cloudConfFields struct {
		authURL                     string
		applicationCredentialID     string
		applicationCredentialName   string
		applicationCredentialSecret string
		username                    string
		password                    string
		projectID                   string
		projectName                 string
		userDomainName              string
		projectDomainName           string
		regionName                  string
		interfaceName               string
		floatingNetworkID           string
		publicNetworkName           string
	}

	cloudsYaml struct {
		Clouds map[string]cloud `yaml:"clouds"`
	}

	cloud struct {
		Auth       auth   `yaml:"auth"`
		RegionName string `yaml:"region_name"`
		Interface  string `yaml:"interface"`
	}

	auth struct {
		AuthURL                     string `yaml:"auth_url"`
		ApplicationCredentialID     string `yaml:"application_credential_id"`
		ApplicationCredentialName   string `yaml:"application_credential_name"`
		ApplicationCredentialSecret string `yaml:"application_credential_secret"`
		Username                    string `yaml:"username"`
		Password                    string `yaml:"password"`
		ProjectID                   string `yaml:"project_id"`
		ProjectName                 string `yaml:"project_name"`
		UserDomainName              string `yaml:"user_domain_name"`
		ProjectDomainName           string `yaml:"project_domain_name"`
	}
)
```

There is no need for exported types and respective for fields if applicable

File: internal/credspropagation/openstack.go [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+		Namespace: cfg.ManagedCluster.Namespace,
+	}, openstackCluster)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return fmt.Errorf("OpenStackCluster %s/%s not found: %w",
+				cfg.ManagedCluster.Namespace, cfg.ManagedCluster.Name, err)
+		}
+		return fmt.Errorf("unable to get OpenStackCluster %s/%s: %w",
+			cfg.ManagedCluster.Namespace, cfg.ManagedCluster.Name, err)
+	}
```
Comment (zerospiel) [inline internal/credspropagation/openstack.go:89]:
```suggestion
	if err := cfg.Client.Get(ctx, client.ObjectKey{
		Name:      cfg.ManagedCluster.Name,
		Namespace: cfg.ManagedCluster.Namespace,
	}, openstackCluster); err != nil {
		return fmt.Errorf("unable to get OpenStackCluster %s/%s: %w",
			cfg.ManagedCluster.Namespace, cfg.ManagedCluster.Name, err)
	}
```

I am not sure we will benefit from duplicating the not found error (`.. not found: ... not found`)

File: internal/credspropagation/openstack.go [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+
+	// Fetch the OpenStack secret
+	openstackSecret, err := fetchOpenStackSecret(ctx, cfg, credential)
+	if err != nil {
+		return fmt.Errorf("failed to fetch OpenStack secret: %w", err)
+	}
+
+	// Generate the CCM secret using the extracted cloudName
+	ccmSecret, err := generateOpenStackCCMSecret(ctx, cfg, openstackSecret, cloudName)
```
Comment (zerospiel) [inline internal/credspropagation/openstack.go:104]:
```suggestion
	ccmSecret, err := generateOpenStackCCMSecret(openstackCluster, openstackSecret, cloudName)
```

File: internal/credspropagation/openstack.go [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+		Namespace: credential.Spec.IdentityRef.Namespace,
+	}, openstackSecret); err != nil {
+		return nil, fmt.Errorf("failed to get OpenStack secret %s/%s: %w",
+			credential.Spec.IdentityRef.Namespace, credential.Spec.IdentityRef.Name, err)
+	}
+	return openstackSecret, nil
+}
+
+// Generate the CCM secret from the OpenStack secret
+func generateOpenStackCCMSecret(ctx context.Context, cfg *PropagationCfg, openstackSecret *corev1.Secret, cloudName string) (*corev1.Secret, error) {
```
Comment (zerospiel) [inline internal/credspropagation/openstack.go:135]:
```suggestion
func generateOpenStackCCMSecret(osCluster *capo.OpenStackCluster, openstackSecret *corev1.Secret, cloudName string) (*corev1.Secret, error) {
```

File: internal/credspropagation/openstack.go [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+	fields, err := extractCloudConfFields(parsedCloudsYaml, cloudName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract cloud.conf fields: %w", err)
+	}
+
+	// Fetch external network details from OpenStackCluster
+	externalNetwork, err := fetchExternalNetwork(ctx, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch external network details: %w", err)
+	}
```
Comment (zerospiel) [inline internal/credspropagation/openstack.go:187]:
```suggestion
	externalNetwork := osCluster.Status.ExternalNetwork
```

File: internal/credspropagation/openstack.go [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+	fields.FloatingNetworkID = externalNetwork.ID
+	fields.PublicNetworkName = externalNetwork.Name
+
+	// Render the cloud.conf secret
+	return renderCloudConf(cloudConfTemplate, fields)
+}
+
+// Parse the clouds.yaml content into structured types
+func parseCloudsYaml(data []byte) (*CloudsYaml, error) {
+	var cloudsYaml CloudsYaml
```
Comment (zerospiel) [inline internal/credspropagation/openstack.go:203]:
```suggestion
	var cy cloudsYaml
```

Avoid shadowing (lint after the unexporting rename)

File: internal/credspropagation/openstack.go [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, fmt.Errorf("OpenStackCluster %s/%s not found",
+				cfg.ManagedCluster.Namespace, cfg.ManagedCluster.Name)
+		}
+		return nil, fmt.Errorf("unable to get OpenStackCluster %s/%s: %w",
+			cfg.ManagedCluster.Namespace, cfg.ManagedCluster.Name, err)
+	}
+	return openstackCluster.Status.ExternalNetwork, nil
+}
```
Comment (zerospiel) [inline internal/credspropagation/openstack.go:254]: Could be removed, no need in fetching the same object within the same scope

File: internal/controller/managedcluster_controller.go [hunk 14]
```diff
@@ -761,6 +761,25 @@ func (r *ManagedClusterReconciler) reconcileCredentialPropagation(ctx context.Co
 				Reason:  hmc.SucceededReason,
 				Message: "vSphere CCM credentials created",
 			})
+		case "openstack":
+			l.Info("OpenStack creds propagation start")
+			if err := credspropagation.PropagateOpenStackSecrets(ctx, propnCfg, credential); err != nil {
```
Comment (zerospiel) [inline internal/controller/managedcluster_controller.go:766]: nit: the `".../credspropagation".PropagationCfg` symbol could be extended with the `credential`'s `Spec.IdentityRef` hence the signature of the `PropagateOpenStackSecrets` will be along with the rest of propagation functions. Up to the implementation and the code owning developer though

**Review by zerospiel — COMMENTED** [review]
A couple of comments for network, reading, and memory simplifications

File: config/dev/openstack-credentials.yaml [hunk 15]
```diff
@@ -0,0 +1,31 @@
+---
```
Comment (zerospiel) [inline config/dev/openstack-credentials.yaml:1]: Forgot to ask: do we need a doc on how to setup all of the envs here? I might have missed it in other PRs

File: internal/credspropagation/openstack.go (hunk 7 above)
Comment (zerospiel) [reply internal/credspropagation/openstack.go:72]: Note: the `cloudConfFields` type changes will also require the regarding changes to the template itself

File: internal/credspropagation/openstack.go [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	capo "sigs.k8s.io/cluster-api-provider-openstack/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	hmc "github.com/Mirantis/hmc/api/v1alpha1"
+)
+
+// Fields needed to populate the cloud.conf template
+type CloudConfFields struct {
```
Comment (eromanova) [inline internal/credspropagation/openstack.go:34]: Can we use `AuthOptions` struct available in https://github.com/gophercloud/gophercloud/blob/v2.4.0/auth_options.go instead? Just a thought to consider.

File: internal/controller/managedcluster_controller.go (hunk 14 above)
Comment (bnallapeta) [reply internal/controller/managedcluster_controller.go:766]: Hmm, I am okay either ways. I guess we should do that (include identityRef into PropagationCfg struct) if it will be helpful for other providers in the future.
@eromanova thoughts? If you think its required, I can do that change in the same PR.

File: internal/credspropagation/openstack.go (hunk 16 above)
Comment (bnallapeta) [reply internal/credspropagation/openstack.go:34]: I did see this before. But it has only the auth fields. We would need to do a mix of importing this from gopher where maintaining our struct for the rest of the fields. That's why I chose to go with our custom struct so that its clean.

File: internal/credspropagation/openstack.go (hunk 11 above)
Comment (bnallapeta) [reply internal/credspropagation/openstack.go:187]: Fixed.

File: internal/credspropagation/openstack.go (hunk 10 above)
Comment (bnallapeta) [reply internal/credspropagation/openstack.go:135]: Same as above.

File: internal/credspropagation/openstack.go (hunk 9 above)
Comment (bnallapeta) [reply internal/credspropagation/openstack.go:104]: Up until the latest commit, I wasn't fetching OpenStackCluster resource. Since it was not there, I was using ctx and cfg. 

Have removed them now.

File: internal/credspropagation/openstack.go (hunk 8 above)
Comment (bnallapeta) [reply internal/credspropagation/openstack.go:89]: Removed.

File: internal/credspropagation/openstack.go (hunk 7 above)
Comment (bnallapeta) [reply internal/credspropagation/openstack.go:72]: Done.

**Comment by bnallapeta** [comment]
Testing once again before I push the changes.

File: internal/credspropagation/openstack.go (hunk 16 above)
Comment (zerospiel) [reply internal/credspropagation/openstack.go:34]: i believe the API is stable, so supporting the own struct won't hurt in this case, considering introducing a new dependency for a mere struct, so I personally would agree with @bnallapeta on this and leave the struct

Comment (eromanova) [reply internal/credspropagation/openstack.go:34]: I got it. Thank you!

File: internal/controller/managedcluster_controller.go (hunk 14 above)
Comment (eromanova) [reply internal/controller/managedcluster_controller.go:766]: I'd agree with Michael. Let's do it in this PR.

Comment (bnallapeta) [reply internal/controller/managedcluster_controller.go]: Done.

**Comment by bnallapeta** [comment]
Rebased the PR after the ManagedCluster renaming. Let's wait for the CI 🤞

File: internal/controller/clusterdeployment_controller.go [hunk 17]
```diff
@@ -312,8 +312,9 @@ func (r *ClusterDeploymentReconciler) updateCluster(ctx context.Context, mc *hmc
 		},
 		ChartRef: clusterTpl.Status.ChartRef,
 	}
-	if clusterTpl.Spec.Helm.ChartSpec != nil {
-		hrReconcileOpts.ReconcileInterval = &clusterTpl.Spec.Helm.ChartSpec.Interval.Duration
+	reconcileInterval := clusterTpl.Spec.Helm.ChartSpec.Interval.Duration
+	if reconcileInterval != 0 {
+		hrReconcileOpts.ReconcileInterval = &reconcileInterval
```
Comment (eromanova) [inline internal/controller/clusterdeployment_controller.go:317]: Could you please remove this change? It seems to have been unintentionally introduced during the rebase process.

File: internal/controller/clusterdeployment_controller.go [hunk 18]
```diff
@@ -354,11 +355,9 @@ func (r *ClusterDeploymentReconciler) updateCluster(ctx context.Context, mc *hmc
...
 	}
 
-	if mc.Spec.PropagateCredentials {
-		if err := r.reconcileCredentialPropagation(ctx, mc); err != nil {
-			l.Error(err, "failed to reconcile credentials propagation")
-			return ctrl.Result{}, err
-		}
+	if err := r.reconcileCredentialPropagation(ctx, mc, cred); err != nil {
+		l.Error(err, "failed to reconcile credentials propagation")
+		return ctrl.Result{}, err
```
Comment (eromanova) [inline internal/controller/clusterdeployment_controller.go:360]: Same

File: internal/credspropagation/common.go [hunk 19]
```diff
@@ -86,7 +87,11 @@ func makeClientFromSecret(kubeconfSecret *corev1.Secret) (client.Client, error)
...
 		return nil, err
 	}
-	return client.New(restConfig, client.Options{
+	cl, err := client.New(restConfig, client.Options{
 		Scheme: scheme,
 	})
+	if err != nil {
+		return nil, err
+	}
+	return cl, nil
```
Comment (eromanova) [inline internal/credspropagation/common.go:96]: Same

File: internal/credspropagation/openstack.go (hunk 9 above)
Comment (eromanova) [reply internal/credspropagation/openstack.go:91]: I'm not sure it's resolved. @bnallapeta could you please check?

Comment (bnallapeta) [reply internal/credspropagation/openstack.go:91]: Fixed now.

**Review by eromanova — APPROVED** [review]
